datasize_macro/src/impl_data_sizes.rs

```diff
@@ -0,0 +1,175 @@
+					}
+				}
+				// If the variant is a unit variant, then the size is 0
+				Fields::Unit => quote! {
+					Self::#ident => 0,
+				},
+			}
+		})
+		.fold(quote!(), |t, b| quote! (#t #b));
+
+	quote! {
+		match &self {
+			#branches
+		}
+	}
+}
+
+fn impl_datasize_struct(data_struct: &DataStruct) -> TokenStream2 {
+	match &data_struct.fields {
+		Fields::Named(f) => {
+			// Get a list of all named fields of a named variant
+			let names: Vec<Ident> = f.named.iter().filter_map(|f| f.ident.to_owned()).collect();
+
+			quote! {
+				usize::default() #(+ self.#names.data_size())*
+			}
+		}
+		Fields::Unnamed(f) => {
+			// Set a name for all fields of an unnamed variant
+			let names: Vec<Index> = f
+				.unnamed
+				.iter()
+				.enumerate()
+				.map(|(i, _)| Index::from(i))
+				.collect();
+
+			quote! {
+				usize::default() #(+ self.#names.data_size())*
+			}
+		}
+		// If the variant is a unit variant, then the size is 0
+		Fields::Unit => quote!(usize::default()),
+	}
+}
+
+fn impl_static_datasize_enum(data_enum: &DataEnum) -> TokenStream2 {
+	data_enum
+		.variants
+		.iter()
+		.map(|v| {
+			// Retrieve types for all fields
+			let types: Vec<Type> = v.fields.iter().map(|f| f.ty.to_owned()).collect();
+			quote! {
+				usize::default() #( + #types::static_data_size())*
+			}
+		})
+		// Use the maximum size among all variants
+		.fold(
+			quote!(usize::default()),
+			|t, b| quote!(std::cmp::max(#t, #b)),
+		)
+}
+
+fn impl_static_datasize_struct(data_struct: &DataStruct) -> TokenStream2 {
+	// Retrieve types of all fields
+	let types: Vec<TokenStream2> = data_struct
+		.fields
+		.iter()
+		.map(|f| f.ty.to_owned())
+		.map(replace_type_syntax)
+		.collect();
+
+	// We call `static_data_size()` on each of the names
+	quote! ( usize::default() #(+ <#types>::static_data_size())*)
+}
+
+/// This replaces all types to a syntax on which we can call functions
+fn replace_type_syntax(t: Type) -> TokenStream2 {
+	match t {
+		Type::Tuple(t) => {
+			let types: Vec<TokenStream2> = t
+				.elems
+				.iter()
+				.map(|t| replace_type_syntax(t.to_owned()))
+				.collect();
+			quote!((#(#types),*))
+		}
+		Type::Ptr(p) => {
+			let ty = replace_type_syntax(*p.elem);
+			quote!(#ty)
+		}
+		Type::Array(a) => {
+			let ty = replace_type_syntax(*a.elem);
+			quote!([#ty])
+		}
+		Type::Slice(s) => {
+			let ty = replace_type_syntax(*s.elem);
+			quote!([#ty])
+		}
+		Type::Reference(r) => {
+			let ty = replace_type_syntax(*r.elem);
+			quote!(#ty)
+		}
+		Type::Infer(i) => i.to_token_stream(),
+		Type::Path(p) => p.to_token_stream(),
+		_ => panic!("This type is not supported yet"),
+	}
+}
+
+pub fn retrieve_generics(input: &DeriveInput) -> Vec<Ident> {
+	input
+		.generics
+		.type_params()
+		.map(|t| t.ident.to_owned())
+		.collect()
+}
+
+pub fn retrieve_lifetimes(input: &DeriveInput) -> Vec<TokenStream2> {
+	input.generics.lifetimes().map(|_| quote!('_)).collect()
+}
```

Comment:
For generics i thing yes, but for lifetimes we don't need them to be named, so i just anonymize them.
Maybe this could be done like for the names from earlier.

Reply:
There is no problem naming them though.  The compiler treats them just the same.

Reply:
I think `syn::Generics` should be used - the lifetimes won't be anonymous, but that does not matter.  You just push the bounds, then split with
```rust
let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
```
and use with
```rust
impl #impl_generics cornflakes::DataSize for #name #type_generics #where_clause { ... }
```
Should work everywhere just fine, assuming you push the bounds first